### PR TITLE
Introduce a function barrier to mapslices

### PR DESCRIPTION
**Pull request by bramtayl**

### CODE REVIEW

**Comment by bramtayl** [comment]
Significant performance improvements in some cases:


`mapslices(sum, rand(1000, 1000), 1)`

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by bramtayl** [comment]
P.S. if anyone has any other suggestions for performance improvements, let me know. I've tried a handful and this change was the only real success. 

https://discourse.julialang.org/t/bikeshedding-mapslices/3998/10

**Comment by timholy** [comment]
Sorry I forgot to tune back into the discussion. I'll take a peek in the next couple of days; if I find ways to improve it further, I'll let you know.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1829,19 +1829,24 @@ function mapslices(f, A::AbstractArray, dims::AbstractVector)
     R[ridx...] = r1
 
     nidx = length(otherdims)
+    indexes = Iterators.drop(CartesianRange(itershape), 1)
+    inner_mapslices!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R)
+end
+
+@noinline function inner_mapslices!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R)
     if safe_for_reuse
```
Comment (mbauman) [inline base/abstractarray.jl]: It looks like `safe_for_reuse` could now more idiomatically be expressed with dispatch (and two `inner_mapslices!` methods).  Of course the performance impact here is a drop in the ocean… but I think it's a nicer way to express the two algorithms.

File: base/abstractarray.jl [hunk 2]
```diff
@@ -1829,19 +1829,24 @@ function mapslices(f, A::AbstractArray, dims::AbstractVector)
...
         # when f returns an array, R[ridx...] = f(Aslice) line copies elements,
         # so we can reuse Aslice
-        for I in Iterators.drop(CartesianRange(itershape), 1) # skip the first element, we already handled it
+        for I in indexes # skip the first element, we already handled it
             for i in 1:nidx
                 idx[otherdims[i]] = ridx[otherdims[i]] = I.I[i]
             end
-            _unsafe_getindex!(Aslice, A, idx...)
+            Base._unsafe_getindex!(Aslice, A, idx...)
```
Comment (mbauman) [inline base/abstractarray.jl]: Don't need the `Base.`.

**Review by mbauman — COMMENTED** [review]
Seems like a good first step at some of the lowest hanging fruit.

Without profiling, I'd guess getting `(idx...,)` and `(ridx...,)` to infer is going to be your next biggest performance win. Constructing them as tuples directly isn't going to be easy without #21912, but perhaps the easiest bandaid to try is to also pass the resulting types from the splats to `inner_mapslices!`.  As much as it stinks to add yet more arguments to this function, it'd let you use a type assertion:

    t = (idx...,)::TIdx
    _unsafe_getindex!(Aslice, A, t...)

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22229#issuecomment-306260500) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d778f6ba658816cb073c09a9536774e12d540a6a/cbc4fef_vs_d318753/report.md). cc @jrevels

**Comment by bramtayl** [comment]
Hmm, I tried the tuple types trick and it seemed to decrease performance...

```julia
function my_mapslices(f, A::AbstractArray, dims::AbstractVector)
    if isempty(dims)
        return map(f,A)
    end

    dimsA = [indices(A)...]
    ndimsA = ndims(A)
    alldims = [1:ndimsA;]

    otherdims = setdiff(alldims, dims)

    idx = Any[first(ind) for ind in indices(A)]
    itershape   = tuple(dimsA[otherdims]...)
    for d in dims
        idx[d] = Base.Slice(indices(A, d))
    end

    # Apply the function to the first slice in order to determine the next steps
    Aslice = A[idx...]
    r1 = f(Aslice)
    # In some cases, we can re-use the first slice for a dramatic performance
    # increase. The slice itself must be mutable and the result cannot contain
    # any mutable containers. The following errs on the side of being overly
    # strict (#18570 & #21123).
    safe_for_reuse = isa(Aslice, StridedArray) &&
                     (isa(r1, Number) || (isa(r1, AbstractArray) && eltype(r1) <: Number))

    # determine result size and allocate
    Rsize = copy(dimsA)
    # TODO: maybe support removing dimensions
    if !isa(r1, AbstractArray) || ndims(r1) == 0
        r1 = [r1]
    end
    nextra = max(0, length(dims)-ndims(r1))
    if eltype(Rsize) == Int
        Rsize[dims] = [size(r1)..., ntuple(d->1, nextra)...]
    else
        Rsize[dims] = [indices(r1)..., ntuple(d->OneTo(1), nextra)...]
    end
    R = similar(r1, tuple(Rsize...,))

    ridx = Any[map(first, indices(R))...]
    for d in dims
        ridx[d] = indices(R,d)
    end

    R[ridx...] = r1

    nidx = length(otherdims)
    # skip the first element, we already handled it
    indexes = Iterators.drop(CartesianRange(itershape), 1)
    idx_type = typeof((idx...,))
    ridx_type = typeof((ridx...,))
    dispatch_inner_mapslices!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
end

@noinline function dispatch_inner_mapslices!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
    if safe_for_reuse
        reuse_inner_mapslices!(indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
     else
        new_inner_mapslices!(indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
    end
end

function reuse_inner_mapslices!(indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
    for I in indexes
        idx_tuple, ridx_tuple = replace_mapslice_indices!(nidx, idx, otherdims, ridx, I.I, idx_type, ridx_type)
        R[ridx_tuple...] = f(A[idx_tuple...])
    end
    R
end

function new_inner_mapslices!(indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, idx_type, ridx_type)
    # when f returns an array, R[ridx...] = f(Aslice) line copies elements,
    # so we can reuse Aslice
    for I in indexes 
        idx_tuple, ridx_tuple = replace_mapslice_indices!(nidx, idx, otherdims, ridx, I.I, idx_type, ridx_type)
        Base._unsafe_getindex!(Aslice, A, idx_tuple...)
        R[ridx_tuple...] = f(Aslice)
    end
    R
end

function replace_mapslice_indices!(nidx, idx, otherdims, ridx, I, idx_type, ridx_type)
    for i in 1:nidx
        idx[otherdims[i]] = ridx[otherdims[i]] = I[i]
    end
     (idx...,)::idx_type, (ridx...,)::ridx_type
end
```

**Comment by mbauman** [comment]
Yup, you're right. I had something a bit simpler in mind with that comment, but that seems to be slower, too:

```jl
# …
    inner_mapslices2!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, (idx...,), (ridx...,))
end

@noinline function inner_mapslices2!(safe_for_reuse, indexes, nidx, idx, otherdims, ridx, Aslice, A, f, R, ::Idx, ::RIdx) where {Idx, RIdx}
# …
                   t = (idx...,)::Idx
                   r = (ridx...,)::RIdx
                   Base._unsafe_getindex!(Aslice, A, t...)
                   R[r...] = f(Aslice)
...

julia> @btime Base.mapslices($sum, $X, $dims); # 0.6.0-rc2
  4.494 ms (10524 allocations: 259.98 KiB)

julia> @btime Base.mapslices1($sum, $X, $dims); # This PR
  2.551 ms (4018 allocations: 111.48 KiB)

julia> @btime Base.mapslices2($sum, $X, $dims); # This PR + tuple assertions
  3.610 ms (5508 allocations: 134.80 KiB)
```

Actually profiling reveals that about 2/3rds of the time with this PR is in `_unsafe_getindex!`.  This is a very nice improvement as it stands.

**Comment by bramtayl** [comment]
@timholy did you get a change to look at this?

**Comment by timholy** [comment]
This seems fine. @Cody-G and I are going to start taking a closer look at `mapslices`; our plan is to see if we find any further opportunities for improving performance, and if we find them we'll both fix `mapslices` and write it up as a case study for the blog. But no reason this shouldn't be merged in the meantime (just need to address https://github.com/JuliaLang/julia/pull/22229#discussion_r120181778).

File: base/abstractarray.jl (hunk 1 above)
Comment (bramtayl) [inline base/abstractarray.jl]: Ok, well, I did refactor a bit to avoid code duplication but not quite in the way you suggested.

**Comment by bramtayl** [comment]
I'm guessing the time-out on 32 bit Linux is unrelated, so this is probably good to go

**Comment by timholy** [comment]
Thanks, @bramtayl!

**Comment by bramtayl** [comment]
Thanks for all of your help!

**Comment by musm** [comment]
should this be backported?

**Pull request merged**